### PR TITLE
Add support to read JSON credential files present in a directory passed through the `<PROVIDER>_APPLICATION_CREDENTIALS` environment variable

### LINKED ISSUE #729: [BUG] Temporarily enable passing of cloud provider credentials in JSON through the directory pointed by `${PROVIDER}_APPLICATION_CREDENTIALS`
**Describe the bug:**

[Example](https://github.com/gardener/etcd-backup-restore/blob/master/example/storage-provider-secrets/01-aws-s3-storage-secret.yaml) provides 2 options to specify the backup bucket secret. Option 2 (with data as JSON) is not supported today and should be removed.
> NOTE: Please make the correction for all object store providers.

This only creates confusion for adopters. See https://kubernetes.slack.com/archives/CB57N0BFG/p1713778819927999

### PR DESCRIPTION
**Pull request by renormalize**
**What this PR does / why we need it**:

The dual way of passing credentials to etcd-backup-restore, but the file-type agnostic way of passing credentials in [gardener/etcd-druid](https://github.com/gardener/etcd-druid) caused confusion in the community consuming both etcd-druid and etcd-backup-restore.

Passing credentials through a JSON file in etcd-druid does not cause it to error, but is not supported by etcd-druid - and thereby causes etcd-backup-restore to error as a result, even though etcd-backup-restore supports credentials in JSON files. 
If consumers familiar with etcd-backup-restore who pass credentials as JSON typically, use the same method to pass credentials in etcd-druid, etcd-backup-restore errors.

To temporarily solve this, if credentials are passed through a JSON file in etcd-druid (i.e. as a file in a directory whose path is exported in `<PROVIDER>_APPLICATION_CREDENTIALS`), etcd-backup-restore handles this and fetches the credentials from the JSON file. If a JSON file is present in the directory, all other files are ignored.

However, as mentioned in #729, credentials will only be passed to etcd-backup-restore as individual files in a directory in 3 releases from the next (as of writing, v0.31.0), and credentials in the form of JSON files will be deprecated.

This PR makes the following changes:

* Add support to read JSON credential files present in a directory which is passed through the `<PROVIDER>_APPLICATION_CREDENTIALS` environment variable for all (applicable) providers.

* Remove examples which suggest that cloud provider credentials can be passed through a JSON file since this will be deprecated in future releases.

* Improve error handling in various `SnapStore` related files for each cloud provider.

* Updated docs to discourage usage of JSON files to pass credentials.

* Housekeeping in `pkg/server/backuprestoreserver.go`. Changed an unused named function parameter to `_`, and removed a tautological if condition.

Tested with the following cloud providers to test the feature and check for regressions:

- [x] AWS S3
- [x] Azure Blob Storage
- [x] GCS (currently only a `service_account.json` is accepted so this issue will not be faced.)
- [x] OpenStack Swift

**Which issue(s) this PR fixes**:
Fixes #729 

**Special notes for your reviewer**:

Please update the block header and the release note to something that you feel is more appropriate.

**Release note**:

```noteworthy|action user|operator
Passing object store credentials to etcd-backup-restore in the form of a JSON file is now deprecated. Support to pass JSON credentials via file mount (File Path Set through Env: <ProviderName>_APPLICATION_CREDENTIALS_JSON) will be dropped by v0.31.0.
```

```noteworthy|action user|operator
Please switch to passing credentials through non-JSON file as mentioned in https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets and Set File Path through Env: `<ProviderName>_APPLICATION_CREDENTIALS`
```

### CODE REVIEW

File: pkg/snapstore/abs_snapstore.go
Comment (anveshreddy18) [inline pkg/snapstore/abs_snapstore.go:135]: Can I know why can't we just club these two `if` statements together, since then there would be no code redundancy and also it will be easy to understand the flow of fetching the credentials. As the flow goes like 

```
See if `absCredentialDirectory` is set, if yes -> Check if `jsonCredentialFile` is present, if yes, read from there, if no -> proceed to read the credentials from individual files
```

I presume you wanted to keep it this way since it's a temporary change and will be reverted back in a few releases but I think 3 releases is long enough time considering there are security issues and some major upgrades going on due to which the releases won't be as frequent. I believe even if it's temporary, it will be a bit easier for the users to understand the change in a neat flow. WDYT?

File: pkg/snapstore/abs_snapstore.go
Comment (anveshreddy18) [inline pkg/snapstore/abs_snapstore.go:408]: Same as . And for all other providers.

**Review by anveshreddy18 — CHANGES_REQUESTED** [review]
Thanks for the PR @renormalize. I left a couple of comments. PTAL

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:135]: I don't have a strong preference for either, but I feel this coding style is easier to read. Seperating this handling into a different `if` makes the intent clear that this handling is sepearate from the other two, i.e. an edge case that is being handled. If you strongly feel clubbing makes the code look cleaner, I will make the necessary changes.

Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:135]: I don't have a strong opinion on this, but I personally feel that this is easier to read, than having a huge blob in an `if`. A separate `if` for this case makes the intent clear that the case being handled is different from the other two, because it is different from what is typically understood of how credentials are sent to etcd-backup-restore.

If you have a strong preference for this code in the same `if` condition, I'll make the changes accordingly.

**Comment by renormalize** [comment]
Would 3 releases be necessary for deprecation @gardener/etcd-druid-maintainers? We're still in a 0 major version and given the release cadence of etcd-backup-restore, this will take a *long* time as @anveshreddy18 has mentioned in a review comment.

Being in "alpha" gives us the freedom to deprecate features without any repercussions. However, I do understand that etcd-backup-restore has been used by a lot of members who have come to rely on their particular workflows with regards to passing credentials, and 1 release is not enough time. Would 2 be good enough?

File: docs/deployment/getting_started.md
Comment (ishan16696) [inline docs/deployment/getting_started.md:11]: > however, it is *highly* recommended passing each credential as an individual file in a directory.

I got your point here, but IMO this could be little ambiguous for community user: "passing each credential as an individual file".  Actually, user is passing secret in a 1 file but secrets are getting mounted as individual file. IMO, previous statement is correct. If you want it to be more accurate then mention the [examples of storage-provider-secrets](https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets).

File: example/storage-provider-secrets/01-aws-s3-storage-secret.yaml
Comment (ishan16696) [inline example/storage-provider-secrets/01-aws-s3-storage-secret.yaml:16]: I would suggest if you are removing it from examples then you can remove it from doc as well: https://github.com/gardener/etcd-backup-restore/blob/master/docs/deployment/getting_started.md#various-ways-to-pass-credentials for all storage providers.

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:134]: I would suggest put this block of code before checking the `os.LookupEnv(prefixString + absCredentialJSONFile)`. This will save you from extra part of code.
```
See if `absCredentialDirectory` is set, if yes -> Check if `jsonCredentialFile` is present, if yes --> set the ENV `absCredentialJSONFile` and continue the code as it is
```
By this you don't have to handle to write piece of code:
```
		if jsonCredentialFile != "" {
			credentials, err := readABSCredentialsJSON(jsonCredentialFile)
			if err != nil {
				return "", "", fmt.Errorf("error getting credentials using %v JSON file in a directory with error: %w", jsonCredentialFile, err)
			}
			return credentials.StorageAccount, credentials.SecretKey, nil
		}
```

File: pkg/snapstore/s3_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/s3_snapstore.go:571]: If you follow my above suggestion then here you don't need extra handling of this, may be you need to just swap the code .... check `JSON` first then `non-JSON`

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:134]: I'd specifically not chosen to do this because the environment variable is set by the consumer of etcd-backup-restore.

etcd-backup-restore changing the environment variables that it consumes from the user as a part of configuration does not feel right to me.

Let's say that a consumer of etcd-druid provides a JSON file as a part of the credential secret. During a credential rotation, they choose to provide the credentials as normal data fields which are mounted as individual files in the etcd-backup-restore container. If etcd-backup-restore sets the JSON environment variable, it will not be unset when this credential rotation happens, and the credentials will still be looked for as a JSON file.

Due to configuration to the binary being changed by the binary causing possible confusion, and the above case, I wrote the code as it is right now.

If you still feel the flow would be made simpler due to the suggestion, I will change the code.

File: docs/deployment/getting_started.md
Comment (renormalize) [reply docs/deployment/getting_started.md:11]: You are right. I will make the corresponding changes by pointing the user to the example secret files.
Thanks for the suggestion!

File: example/storage-provider-secrets/01-aws-s3-storage-secret.yaml
Comment (renormalize) [reply example/storage-provider-secrets/01-aws-s3-storage-secret.yaml:16]: Good catch! Thanks for the suggestion.

I'll remove it from the mentioned doc as well.

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [reply pkg/snapstore/abs_snapstore.go:134]: good edge catch but can be easily handled by this:
```
See if `absCredentialDirectory` is set, if yes -> Check if `jsonCredentialFile` is present, if yes --> set the ENV `absCredentialJSONFile` and continue the code as it is else unset ENV for JSON.
```
And secret rotation detection shouldn't depend on any `ENV`, it should list all files and check if any file get updated or not.

Comment (ishan16696) [reply pkg/snapstore/abs_snapstore.go:134]: But I leave this decision to you, basically my point was to make less code change.

Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:134]: Alright, I'll take a call on this. I don't mind the larger change since it's easier to follow.

File: pkg/snapstore/utils.go
Comment (ishan16696) [inline pkg/snapstore/utils.go:236]: I don't understand the point of taking `providerErrorString` as an argument as this function `getJSONCredentialModifiedTime` doesn't look like it is using this argument. It just using it in error string which I guess caller can do it.

**Review by ishan16696 — CHANGES_REQUESTED** [review]

File: pkg/snapstore/utils.go
Comment (renormalize) [reply pkg/snapstore/utils.go:236]: Alright, will make that change! Thanks for the suggestion.

File: pkg/snapstore/abs_snapstore.go
Comment (anveshreddy18) [reply pkg/snapstore/abs_snapstore.go:135]: Since I don't have strong opinion on this and the below comment, I leave it to you to decide. You can close or address this depending on your opinion when you come to decide on this section of code since Ishan also has some reviews on this piece of code.

Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:135]: I've expressed the same opinion with Ishan as well and he's okay with it. I'll be keeping this section as is. Thanks for the review!

File: pkg/snapstore/utils.go
Comment (renormalize) [reply pkg/snapstore/utils.go]: [81620c7](https://github.com/gardener/etcd-backup-restore/pull/730/commits/81620c7e6af279c5a0e631b8fa6dfa2140f58b3f).

**Comment by renormalize** [comment]
I've addressed your comments @ishan16696, feel free to let me know if you have any more suggestions!

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:118]: I would suggest open a issue to drop the support for secret passing via JSON credential file, so that it become easy to track

**Review by ishan16696 — COMMENTED** [review]
Overall looks good. just one nit

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [reply pkg/snapstore/abs_snapstore.go:118]: assign that issue a milestone of `v0.31.0`

Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:118]: Yeah will be doing that. Thanks for the suggestion!

Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:118]: I've created https://github.com/gardener/etcd-backup-restore/issues/741 to track this.

**Review by ishan16696 — APPROVED** [review]
LGTM!!

**Review by anveshreddy18 — APPROVED** [review]
LGTM!! Thanks for the PR. 

Before merging, pls make sure to test with OpenStack Swift as well. Thanks

**Pull request merged**
